src/expr/parser.rs

```diff
@@ -8,51 +8,62 @@ use nom::digit;
 use std::str;
 use std::str::FromStr;
 
-named!(parens<i64>, delimited!(
+named!(parens<Expr>, delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
+named!(i64_literal<Expr>,
+  map (
     map_res!(
-      digit,
-      str::from_utf8
+      map_res!(
+        digit,
+        str::from_utf8
+      ),
+      FromStr::from_str
     ),
-    FromStr::from_str
+    Expr::Literal
   )
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
     i64_literal
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 map!(preceded!(tag!("*"), factor), |mul| acc =
+                   Expr::BinOp(Box::new(mem::repalce(&mut acc, Expr::Literal(0))), BinOp::Times,
```

Comment:
You have the main idea here, but I think you have another typo, and perhaps a few more `borrowck` details to work out here.